grovedb/src/replication/state_sync_session.rs

```diff
@@ -0,0 +1,459 @@
+                }
+                Err(_) => {
+                    return Err(Error::CorruptedData(
+                        "Unable to decode incoming chunk".to_string(),
+                    ));
+                }
+            }
+        }
+
+        Ok(res)
+    }
+}
+
+impl<'tx> SubtreeStateSyncInfo<'tx> {
+    pub fn new(restorer: Restorer<PrefixedRocksDbImmediateStorageContext<'tx>>) -> Self {
+        SubtreeStateSyncInfo {
+            restorer,
+            root_key: None,
+            is_sum_tree: false,
+            pending_chunks: Default::default(),
+            current_path: vec![],
+            num_processed_chunks: 0,
+        }
+    }
+}
+
+// Struct governing state sync
+pub struct MultiStateSyncSession<'db> {
+    // Map of current processing subtrees
+    // SubtreePrefix (Path digest) -> SubtreeStateSyncInfo
+    current_prefixes: BTreeMap<SubtreePrefix, SubtreeStateSyncInfo<'db>>,
+    // Set of processed prefixes (Path digests)
+    processed_prefixes: BTreeSet<SubtreePrefix>,
+    // Root app_hash
+    app_hash: [u8; 32],
+    // Version of state sync protocol,
+    pub(crate) version: u16,
+    // Transaction goes last to be dropped last as well
+    transaction: Transaction<'db>,
+    _pin: PhantomPinned,
+}
+
+impl<'db> MultiStateSyncSession<'db> {
+    /// Initializes a new state sync session.
+    pub fn new(transaction: Transaction<'db>, app_hash: [u8; 32]) -> Pin<Box<Self>> {
+        Box::pin(MultiStateSyncSession {
+            transaction,
+            current_prefixes: Default::default(),
+            processed_prefixes: Default::default(),
+            app_hash,
+            version: CURRENT_STATE_SYNC_VERSION,
+            _pin: PhantomPinned,
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.current_prefixes.is_empty()
+    }
+
+    pub fn is_sync_completed(&self) -> bool {
+        for (_, subtree_state_info) in self.current_prefixes.iter() {
+            if !subtree_state_info.pending_chunks.is_empty() {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    pub fn into_transaction(self: Pin<Box<Self>>) -> Transaction<'db> {
+        // SAFETY: the struct isn't used anymore and no one will refer to transaction
+        // address again
+        unsafe { Pin::into_inner_unchecked(self) }.transaction
+    }
+
+    pub fn add_subtree_sync_info<'b, B: AsRef<[u8]>>(
+        self: &mut Pin<Box<MultiStateSyncSession<'db>>>,
+        db: &'db GroveDb,
+        path: SubtreePath<'b, B>,
+        hash: CryptoHash,
+        actual_hash: Option<CryptoHash>,
+        chunk_prefix: [u8; 32],
+        grove_version: &GroveVersion,
+    ) -> Result<Vec<u8>, Error> {
+        // SAFETY: we get an immutable reference of a transaction that stays behind
+        // `Pin` so this reference shall remain valid for the whole session
+        // object lifetime.
+        let transaction_ref: &'db Transaction<'db> = unsafe {
+            let tx: &mut Transaction<'db> =
```

Comment:
As far as I understand `SubtreeStateSyncInfo` instances have _shared_ reference to `Transaction` which is fine, however, for this brief moment we have `&mut Transaction` there in the same time, which is not fine and it violates Rust borrowing rules

Reply:
So what do you suggest then?

Reply:
I think taking `&` reference there instead of `&mut` there and do trick with `&*(.. as *const _)` to prolong lifetime should work

Reply:
I propose to address this suggestion on a separate PR.

Reply:
this introduces UB, we can't merge it like that

Have you tried this?

```
            let tx: &Transaction<'db> = &self.as_ref().transaction;
            &*(tx as *const _)
```

Reply:
done